src/py21cmfast/src/ps.c

```diff
@@ -1555,35 +1568,122 @@ int ComputeLF(int nbins, struct UserParams *user_params, struct CosmoParams *cos
         lnMhalo_hi = log(Mhalo_max);
         dlnM = (lnMhalo_hi - lnMhalo_lo)/(double)(nbins - 1);
 
-        for (i=0; i<nbins; i++) {
-            // calculate luminosity function
-            lnMhalo_i = lnMhalo_lo + dlnM*(double)i;
-            Mhalo_param[i] = exp(lnMhalo_i);
-
-            M_h_z[i + i_z*nbins] = Mhalo_param[i];
-
-            Muv_1 = gsl_spline_eval(LF_spline, lnMhalo_i - delta_lnMhalo, LF_spline_acc);
-            Muv_2 = gsl_spline_eval(LF_spline, lnMhalo_i + delta_lnMhalo, LF_spline_acc);
-
-            dMuvdMhalo = (Muv_2 - Muv_1) / (2.*delta_lnMhalo * exp(lnMhalo_i));
-
-            if(user_params_ps->HMF==0) {
-                log10phi[i + i_z*nbins] = log10( dNdM(z_LF[i_z], exp(lnMhalo_i)) * exp(-(astro_params->M_TURN/Mhalo_param[i])) / fabs(dMuvdMhalo) );
-            }
-            else if(user_params_ps->HMF==1) {
-                log10phi[i + i_z*nbins] = log10( dNdM_st_interp(growthf, exp(lnMhalo_i)) * exp(-(astro_params->M_TURN/Mhalo_param[i])) / fabs(dMuvdMhalo) );
-            }
-            else if(user_params_ps->HMF==2) {
-                log10phi[i + i_z*nbins] = log10( dNdM_WatsonFOF(growthf, exp(lnMhalo_i)) * exp(-(astro_params->M_TURN/Mhalo_param[i])) / fabs(dMuvdMhalo) );
-            }
-            else if(user_params_ps->HMF==3) {
-                log10phi[i + i_z*nbins] = log10( dNdM_WatsonFOF_z(z_LF[i_z], growthf, exp(lnMhalo_i)) * exp(-(astro_params->M_TURN/Mhalo_param[i])) / fabs(dMuvdMhalo) );
-            }else{
-                LOG_ERROR("HMF should be between 0-3... returning error.");
-                return(2);
-            }
-            if (isinf(log10phi[i + i_z*nbins]) || isnan(log10phi[i + i_z*nbins]) || log10phi[i + i_z*nbins] < -30.) log10phi[i + i_z*nbins] = -30.;
-        }
+        // There is a kink on LFs at which Fstar crosses unity. This kink is a numerical artefact caused by the derivate of dMuvdMhalo. 
+		// Most of the cases the kink doesn't appear in magnitude ranges we are interested (e.g. -22 < Muv < -10). However, for some extreme 
+		// parameters, it appears. To avoid this kink, we use the interpolation of the derivate in the range where the kink appears. 
+		// 'i_unity' is the array number at which the kink appears. 'i_unity-3' and 'i_unity+12' are related to the range of interpolation, 
+		// which is an arbitrary choice. 
+		// NOTE: This method does NOT work in cases with ALPHA_STAR < -0.5. But, this parameter range is unphysical given that the 
```

Comment:
Since we know that `ALPHA_STAR<-0.5` is going to produce weird results, we should probably raise at least a warning here (you can do that with `LOG_WARNING`).